File: backend/internal/extract/service.go
package extract

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/google/uuid"

	"github.com/exolutionza/slipscan/backend/internal/ocr"
)

const (
	geminiProvider    = "google"
	geminiDisplayName = "Gemini 2.5 Flash (extraction)"
)

// Service orchestrates the full extraction pipeline:
//  1. Fetch the document row.
//  2. Transition status pending → processing.
//  3. Detect kind (if unknown) with a cheap classify call.
//  4. Run the kind-specific extraction prompt.
//  5. Persist document_extractions, ai_runs; update documents.
type Service struct {
	store   *Store
	ocr     *ocr.Client
	storage StorageGetter
}

// StorageGetter can fetch raw file bytes by object key.
// Satisfied by *storage.Client so we don't import the storage package here.
type StorageGetter interface {
	Get(ctx context.Context, key string) ([]byte, error)
}

// NewService creates an extraction Service.
func NewService(store *Store, ocrClient *ocr.Client, storage StorageGetter) *Service {
	return &Service{store: store, ocr: ocrClient, storage: storage}
}

// Run executes the full extraction pipeline for one document.
// It is safe to call multiple times (re-run). On re-run the old extraction
// row stays (is_current flipped to false) and a new row is written.
func (s *Service) Run(ctx context.Context, docID, orgID uuid.UUID) error {
	start := time.Now()

	// 1. Fetch document.
	doc, err := s.store.GetDocument(ctx, docID, orgID)
	if err != nil {
		if err.Error() == "document not found" {
			return errDocNotFound
		}
		return fmt.Errorf("get document: %w", err)
	}

	// 2. Org default currency (fallback for ambiguous symbols).
	orgCurrency, _ := s.store.OrgCurrency(ctx, orgID)

	// 3. Ensure the AI model row exists.
	modelUUID, err := s.store.EnsureAIModel(ctx, geminiProvider, s.ocr.Model(), geminiDisplayName)
	if err != nil {
		return fmt.Errorf("ensure ai model: %w", err)
	}

	// Determine current document kind (may still be unknown).
	kind := DocumentKind(doc.Kind)

	// 4. Transition document to processing.
	if err := s.store.SetDocumentStatus(ctx, docID, StatusProcessing, kind, ""); err != nil {
		return fmt.Errorf("set processing: %w", err)
	}

	// 5. If kind is unknown, run cheap kind-detection first.
	//    We need the file bytes for this anyway so fetch them now.
	mime := "image/jpeg"
	if doc.MimeType.Valid && doc.MimeType.String != "" {
		mime = doc.MimeType.String
	}

	fileBytes, err := s.storage.Get(ctx, doc.StorageURL)
	if err != nil {
		_ = s.store.SetDocumentStatus(ctx, docID, StatusFailed, kind, "storage fetch failed: "+err.Error())
		return fmt.Errorf("fetch file: %w", err)
	}

	promptVersion := s.promptVersionFor(kind)
	aiRunID, err := s.store.CreateAIRun(ctx, orgID, modelUUID, docID, promptVersion)
	if err != nil {
		return fmt.Errorf("create ai_run: %w", err)
	}

	extractionID, err := s.store.CreateExtraction(ctx, docID, orgID, aiRunID, modelUUID)
	if err != nil {
		return fmt.Errorf("create extraction row: %w", err)
	}

	if kind == KindUnknown || kind == "" {
		detected, detectErr := s.detectKind(ctx, fileBytes, mime)
		if detectErr != nil {
			log.Printf("extract: kind detection failed (docID=%s): %v — defaulting to slip", docID, detectErr)
			detected = KindSlip
		}
		kind = detected
		// Update promptVersion now that we know the kind.
		promptVersion = s.promptVersionFor(kind)
	}

	// 6. Run the kind-specific extraction.
	rawJSON, extractedResult, extractErr := s.runExtraction(ctx, fileBytes, mime, kind, orgCurrency)

	latencyMS := int(time.Since(start).Milliseconds())

	if extractErr != nil {
		_ = s.store.FinishAIRun(ctx, aiRunID, "failed", latencyMS, 0, 0, extractErr.Error())
		_ = s.store.FailExtraction(ctx, extractionID, docID, rawJSON, extractErr.Error())
		return fmt.Errorf("extraction failed: %w", extractErr)
	}

	// 7. Persist successful result.
	if err := s.store.FinishAIRun(ctx, aiRunID, "succeeded", latencyMS, 0, 0, ""); err != nil {
		log.Printf("extract: finish ai_run: %v", err)
	}
	if err := s.store.CompleteExtraction(ctx, extractionID, docID, extractedResult, kind); err != nil {
		return fmt.Errorf("complete extraction: %w", err)
	}

	log.Printf("extract: doc=%s kind=%s confidence=%.2f latency=%dms",
		docID, kind, extractedResult.Confidence, latencyMS)
	return nil
}

// detectKind runs a lightweight Gemini call to classify an unknown document.
func (s *Service) detectKind(ctx context.Context, fileBytes []byte, mime string) (DocumentKind, error) {
	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
	defer cancel()

	raw, err := s.ocr.ExtractWithSchema(ctx, fileBytes, mime, kindDetectPrompt, kindDetectSchema)
	if err != nil {
		return KindUnknown, err
	}
	var kd geminiKind
	if err := json.Unmarshal(raw, &kd); err != nil {
		return KindUnknown, fmt.Errorf("parse kind response: %w", err)
	}
	switch kd.Kind {
	case "slip":
		return KindSlip, nil
	case "invoice":
		return KindInvoice, nil
	case "bank_statement":
		return KindBankStatement, nil
	default:
		return KindSlip, nil
	}
}

// runExtraction calls Gemini with the kind-specific prompt and maps the
// response to the canonical Extracted struct.
func (s *Service) runExtraction(
	ctx context.Context,
	fileBytes []byte,
	mime string,
	kind DocumentKind,
	orgCurrency string,
) (json.RawMessage, *Extracted, error) {
	ctx, cancel := context.WithTimeout(ctx, 45*time.Second)
	defer cancel()

	prompt, schema := s.promptSchemaFor(kind)
	raw, err := s.ocr.ExtractWithSchema(ctx, fileBytes, mime, prompt, schema)
	if err != nil {
		return raw, nil, err
	}

	var gr geminiRaw
	if err := json.Unmarshal(raw, &gr); err != nil {
		return raw, nil, fmt.Errorf("parse extraction response: %w", err)
	}

	result := mapToExtracted(kind, &gr, orgCurrency)
	return raw, result, nil
}

func (s *Service) promptVersionFor(kind DocumentKind) string {
	switch kind {
	case KindSlip:
		return PromptVersionSlip
	case KindInvoice:
		return PromptVersionInvoice
	case KindBankStatement:
		return PromptVersionStatement
	default:
		return PromptVersionKindDetect
	}
}

func (s *Service) promptSchemaFor(kind DocumentKind) (string, map[string]any) {
	switch kind {
	case KindInvoice:
		return invoicePrompt, invoiceSchema
	case KindBankStatement:
		return statementPrompt, statementSchema
	default:
		return slipPrompt, slipSchema
	}
}

File: worker/src/modules/extract/service.ts
/**
 * Extraction service — port of backend/internal/extract/service.go.
 *
 * Pipeline:
 *  1. Fetch the document row (org-filtered).
 *  2. Transition status → processing.
 *  3. Detect kind (slip/invoice/bank_statement) if unknown via a cheap Gemini call.
 *  4. Fetch the file bytes from R2.
 *  5. Run the kind-specific extraction prompt+schema via Gemini.extractWithSchema.
 *  6. INSERT document_extractions + ai_runs rows.
 *  7. Update documents.current_extraction_id + status='extracted'.
 *  On failure, set status='failed' and record the error (matches Go behavior).
 */
import { Gemini } from "../../lib/gemini";
import { getObject } from "../../lib/r2";
import type { Env } from "../../bindings";
import type { DocumentKind, Extracted, GeminiRaw, GeminiKind } from "./types";
import { normalizeCurrency } from "./currency";
import { kindDetectPrompt, kindDetectSchema, promptVersionFor, promptSchemaFor } from "./prompts";
import {
  getDocument,
  getOrgCurrency,
  setDocumentStatus,
  ensureAIModel,
  createAIRun,
  createExtraction,
  finishAIRun,
  completeExtraction,
  failExtraction,
  getExtraction,
} from "./queries";

// ---- helpers ----

function deref(v: number | null | undefined): number {
  return v ?? 0;
}

function derefStr(v: string | null | undefined): string {
  return v ?? "";
}

/**
 * Maps a raw Gemini response to the canonical Extracted struct.
 * Port of Go mapToExtracted.
 */
function mapToExtracted(kind: DocumentKind, raw: GeminiRaw, orgCurrency: string): Extracted {
  const e: Extracted = {
    kind,
    merchant:   derefStr(raw.merchant),
    date:       derefStr(raw.date),
    currency:   normalizeCurrency(derefStr(raw.currency), orgCurrency),
    subtotal:   deref(raw.subtotal),
    tax:        deref(raw.tax),
    total:      deref(raw.total),
    confidence: deref(raw.confidence),
  };

  if (kind === "bank_statement") {
    e.statement_lines = (raw.statement_lines ?? []).map((l) => ({
      date:        derefStr(l.date),
      description: derefStr(l.description),
      amount:      deref(l.amount),
      balance:     deref(l.balance),
    }));
  } else {
    e.line_items = (raw.line_items ?? []).map((l) => ({
      description: derefStr(l.description),
      qty:         deref(l.qty),
      unit:        deref(l.unit),
      amount:      deref(l.amount),
    }));
  }

  return e;
}

// ---- main service function ----

/**
 * Execute the full extraction pipeline for one document.
 * Safe to call multiple times (re-run): old extraction rows are preserved;
 * is_current is flipped to the new row.
 *
 * @returns The UUID of the new document_extractions row on success.
 * @throws  On unrecoverable error (doc not found, R2 failure, DB failure).
 */
export async function extractDocument(
  env: Env,
  orgId: string,
  docId: string,
): Promise<string> {
  const startMs = Date.now();
  const g = new Gemini(env.GEMINI_API_KEY);

  // 1. Fetch document (org-filtered).
  const doc = await getDocument(env, docId, orgId);
  if (!doc) throw Object.assign(new Error("document not found"), { code: "not_found" });

  // 2. Org default currency (fallback for ambiguous symbols).
  const orgCurrency = await getOrgCurrency(env, orgId);

  // 3. Ensure the AI model row exists.
  const modelUUID = await ensureAIModel(env, g.getModel());

  let kind: DocumentKind = (doc.kind as DocumentKind) ?? "unknown";

  // 4. Transition document to 'processing'.
  await setDocumentStatus(env, docId, "processing", kind, "");

  // 5. Fetch file bytes from R2.
  const mime = doc.mime_type && doc.mime_type !== "" ? doc.mime_type : "image/jpeg";
  const fileBytes = await getObject(env, doc.storage_url);
  if (!fileBytes) {
    await setDocumentStatus(env, docId, "failed", kind, "storage fetch failed: object not found");
    throw new Error("fetch file: object not found in R2");
  }

  // 6. Determine prompt version (may update after kind detection).
  let promptVersion = promptVersionFor(kind);

  // 7. Create ai_run + extraction rows upfront (mirrors Go sequencing).
  const aiRunId = await createAIRun(env, orgId, modelUUID, docId, promptVersion);
  const extractionId = await createExtraction(env, docId, orgId, aiRunId, modelUUID);

  // 8. Detect kind if unknown.
  if (kind === "unknown" || kind === ("" as DocumentKind)) {
    try {
      const rawKind = await g.extractWithSchema(fileBytes, mime, kindDetectPrompt, kindDetectSchema);
      const kd = JSON.parse(rawKind) as GeminiKind;
      switch (kd.kind) {
        case "slip":           kind = "slip"; break;
        case "invoice":        kind = "invoice"; break;
        case "bank_statement": kind = "bank_statement"; break;
        default:               kind = "slip"; break;
      }
    } catch (e) {
      console.warn(`extract: kind detection failed (docId=${docId}): ${String(e)} — defaulting to slip`);
      kind = "slip";
    }
    // Update prompt version now that we know the kind.
    promptVersion = promptVersionFor(kind);
  }

  // 9. Run kind-specific extraction.
  const [prompt, schema] = promptSchemaFor(kind);
  let rawJSON: string | null = null;
  let extracted: Extracted | null = null;
  let extractErr: Error | null = null;

  try {
    rawJSON = await g.extractWithSchema(fileBytes, mime, prompt, schema);
    const gr = JSON.parse(rawJSON) as GeminiRaw;
    extracted = mapToExtracted(kind, gr, orgCurrency);
  } catch (e) {
    extractErr = e instanceof Error ? e : new Error(String(e));
  }

  const latencyMs = Date.now() - startMs;

  // 10. Persist result or failure.
  if (extractErr || !extracted) {
    const msg = extractErr?.message ?? "extraction produced no result";
    await finishAIRun(env, aiRunId, "failed", latencyMs, 0, 0, msg);
    await failExtraction(env, orgId, extractionId, docId, rawJSON, msg);
    throw new Error(`extraction failed: ${msg}`);
  }

  await finishAIRun(env, aiRunId, "succeeded", latencyMs, 0, 0, "").catch((e) =>
    console.warn("extract: finish ai_run:", e),
  );
  await completeExtraction(env, orgId, extractionId, docId, extracted, kind);

  console.log(
    `extract: doc=${docId} kind=${kind} confidence=${extracted.confidence.toFixed(2)} latency=${latencyMs}ms`,
  );

  return extractionId;
}

/**
 * Fetch the current extraction row for a document and return it as a plain
 * object suitable for JSON serialisation. Used by the route handler.
 */
export async function getExtractionForDoc(
  env: Env,
  orgId: string,
  extractionId: string,
): Promise<Record<string, unknown> | null> {
  return getExtraction(env, extractionId, orgId);
}
